### PR TITLE
Add a `LightSyncState` field to the chain spec

### PR DESCRIPTION
**Pull request by expenses**
This consists of the first part of implementing https://github.com/paritytech/substrate/issues/6804.

~~You can try it out the RPC call with `curl -H "Content-Type: application/json" -d '{"id":1, "jsonrpc":"2.0", "method": "system_genChainSpec", "params": [true, true]}' http://localhost:9933 | jq '.result .hardcodedSync'`, but to get it to work on Flaming Fir, you might need to switch the light client block import (See https://github.com/paritytech/substrate/issues/6893).~~

### CODE REVIEW

File: client/service/src/lib.rs [hunk 1]
```diff
@@ -332,6 +373,44 @@ async fn build_network_future<
...
+
+	/// We can't fetch a CHT root later than `finalized_number - 2 * cht_size`.
+	let cht_size_x_2 = cht::size::<NumberFor::<TBl>>() * NumberFor::<TBl>::from(2);
+
+	let mut number = NumberFor::<TBl>::one();
+
+	while number <= finalized_number.saturating_sub(cht_size_x_2) {
+		match storage.header_cht_root(cht::size(), number)? {
+			Some(cht_root) => chts.push(cht_root),
+			None => log::error!("No CHT found for block {}", number),
```
Comment (expenses) [inline client/service/src/lib.rs:402]: In theory, `storage.header_cht_root` should never return `None` for this range, but in practise I think it might be better to keep this here instead of panicking.

File: client/chain-spec/src/chain_spec.rs [hunk 2]
```diff
@@ -378,6 +402,49 @@ where
...
+	}
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+#[serde(rename_all = "camelCase")]
+#[serde(deny_unknown_fields)]
+/// The serializable form of `HardcodedSync`. Created using `HardcodedSync::serialize`.
+pub struct SerializableHardcodedSync {
+	header: StorageData,
+	chts: Vec<StorageData>,
```
Comment (expenses) [inline client/chain-spec/src/chain_spec.rs:447]: I believe that `StorageData` is the best way to serialize these, but there could be a better alternative.

File: client/chain-spec/src/chain_spec.rs [hunk 3]
```diff
@@ -378,6 +402,49 @@ where
...
+				.map(|cht| codec::Decode::decode(&mut &cht.0[..]))
+				.collect::<Result<_, _>>()?,
+		})
+	}
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+#[serde(rename_all = "camelCase")]
+#[serde(deny_unknown_fields)]
+/// The serializable form of `HardcodedSync`. Created using `HardcodedSync::serialize`.
```
Comment (bkchr) [inline client/chain-spec/src/chain_spec.rs:444]:
```suggestion
/// The serializable form of `HardcodedSync`. Created using `HardcodedSync::serialize`.
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
#[serde(deny_unknown_fields)]
```

File: client/db/src/light.rs [hunk 4]
```diff
@@ -612,6 +596,48 @@ impl<Block> Storage<Block> for LightStorage<Block>
...
+		&self,
+		cht_size: NumberFor<Block>,
+		block: NumberFor<Block>,
+	) -> ClientResult<Option<Block::Hash>> {
+		self.read_cht_root(CHANGES_TRIE_CHT_PREFIX, cht_size, block)
+	}
+}
+
+
+impl<'a, Block> ChtRootStorage<Block> for &'a LightStorage<Block>
```
Comment (bkchr) [inline client/db/src/light.rs:620]: Why do we need this implementation?

File: client/rpc-api/src/system/mod.rs [hunk 5]
```diff
@@ -103,4 +103,9 @@ pub trait SystemApi<Hash, Number> {
 	/// Returns the roles the node is running as.
 	#[rpc(name = "system_nodeRoles", returns = "Vec<NodeRole>")]
 	fn system_node_roles(&self) -> Receiver<Vec<NodeRole>>;
+
+	/// Returns the json-serialized chainspec running the node.
+	#[rpc(name = "system_genChainSpec", returns = "jsonrpc_core::Value")]
```
Comment (bkchr) [inline client/rpc-api/src/system/mod.rs:108]:
```suggestion
	#[rpc(name = "system_getChainSpec", returns = "jsonrpc_core::Value")]
```

?

File: client/chain-spec/src/chain_spec.rs [hunk 6]
```diff
@@ -378,6 +402,49 @@ where
...
 		self.genesis = GenesisSource::Storage(storage);
 	}
+
+	fn set_hardcoded_sync(&mut self, hardcoded_sync: SerializableHardcodedSync) {
+		ChainSpec::set_hardcoded_sync(self, hardcoded_sync)
+	}
+}
+
+/// Hardcoded infomation that allows light clients to sync quickly.
+pub struct HardcodedSync<Block: BlockT> {
```
Comment (bkchr) [inline client/chain-spec/src/chain_spec.rs:412]: This should be named different.

Maybe `LightSyncState` or similar to make more clear what this is.

File: client/rpc-api/src/system/mod.rs (hunk 5 above)
Comment (expenses) [reply client/rpc-api/src/system/mod.rs:108]: I thought that 'generate' seemed more correct, as we're generating the `HardcodedSync` bit at least.

File: client/chain-spec/src/chain_spec.rs (hunk 6 above)
Comment (expenses) [reply client/chain-spec/src/chain_spec.rs:412]: @danforbes any ideas? :P

File: client/db/src/light.rs (hunk 4 above)
Comment (expenses) [reply client/db/src/light.rs:620]: I thought I needed it but I was wrong.

File: client/chain-spec/src/chain_spec.rs (hunk 6 above)
Comment (danforbes) [reply client/chain-spec/src/chain_spec.rs:412]: Basti's suggestion seems fine to me. I like that it implies up front that the struct is designed for light clients. Could also do something like `InjectableLightClientState`.

File: client/rpc-api/src/system/mod.rs (hunk 5 above)
Comment (danforbes) [reply client/rpc-api/src/system/mod.rs:108]: Spelling out the word `generate` may be an improvement here, pending Basti's response.

Comment (bkchr) [reply client/rpc-api/src/system/mod.rs:108]: Hmm wait, why is this even a RPC call? I assume this is something that we would call on every release to get a new chain spec with the latest light sync state? I don't see this as a RPC. Why isn't that a CLI command?

Comment (expenses) [reply client/rpc-api/src/system/mod.rs:108]: I like the RPC call and I think that's it's nice to have it, but I do think that a CLI command would be more useful for this purpose. Something that starts a light client, waits until it syncs then outputs a chain spec. Unfortunately, for this to be useful #6880 would need to be solved first.

Comment (bkchr) [reply client/rpc-api/src/system/mod.rs:108]: Hmm, I still don't see the value of the RPC call. I first thought that this would be called by a light client, but than I realized that this only works reasonable on light clients. What is the value of the RPC call?

Not sure why #6880 needs to be solved first for this issue? We could just use 2 different base paths.

Another point, why don't we generate the required data for the light sync state on a full node as well?

Comment (expenses) [reply client/rpc-api/src/system/mod.rs:108]: >Another point, why don't we generate the required data for the light sync state on a full node as well?

@cheme @andresilva @svyatonik @rphmeier Is this possible/practical?

Comment (svyatonik) [reply client/rpc-api/src/system/mod.rs:108]: It is possible, but you'd need to build approximately `(BEST_FINALIZED - 4096) / 2048` tries. From practical point of view, since this RPC call is quite specific (iiuc), it may be implemented. The other option would be to build CHTs on full nodes as well (ofc without pruning headers). This will slow down finalization a bit, though.

Comment (expenses) [reply client/rpc-api/src/system/mod.rs:108]: So it sounds like generating this data on the full node will either take up storage space or cpu time. I think it would be better to just do this on the light client for now.

Comment (bkchr) [reply client/rpc-api/src/system/mod.rs:108]: @expenses when do you expect this RPC to be called? And who calls this RPC endpoint?

Comment (expenses) [reply client/rpc-api/src/system/mod.rs]: I've gotten rid of the RPC call now.

File: client/cli/src/commands/export_sync_state_cmd.rs [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+			CL: sp_blockchain::HeaderBackend<B>,
+			BA: MaybeChtRootStorageProvider<B>,
+			SO: sp_consensus::SyncOracle,
+	{
+		futures::future::poll_fn(|_| {
+			match sync_oracle.is_major_syncing() || sync_oracle.is_offline() {
+				true => Poll::Pending,
+				false => Poll::Ready(())
+			}
+		}).await;
```
Comment (seunlanlege) [inline client/cli/src/commands/export_sync_state_cmd.rs:68]: ooh this is pretty clever :grinning:, shouldn't you call the `ctx.wake()` somewhere in the fn though?

Comment (expenses) [reply client/cli/src/commands/export_sync_state_cmd.rs:68]: Not sure. @tomaka should know.

File: client/service/src/chain_ops/export_sync_state.rs [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+	let mut number = NumberFor::<TBl>::one();
+
+	while number <= finalized_number.saturating_sub(cht_size_x_2) {
+		match storage.header_cht_root(cht::size(), number)? {
+			Some(cht_root) => chts.push(cht_root),
+			None => log::error!("No CHT found for block {}", number),
+		}
+
+		number += cht::size();
+	}
```
Comment (seunlanlege) [inline client/service/src/chain_ops/export_sync_state.rs:73]: wait is this enumerating all the blocks up to `finalized_number`? what's the value for `cht::size()`

Comment (expenses) [reply client/service/src/chain_ops/export_sync_state.rs:73]: `cht::size()` is set to `2048`: https://github.com/paritytech/substrate/blob/488b7c7286d4ca9cedd047a5e2c2381736a3f088/client/api/src/cht.rs#L41-L43

So we're enumerating all blocks from 1 to the finalized block - 4096. 1, 2049, etc.

Comment (seunlanlege) [reply client/service/src/chain_ops/export_sync_state.rs:73]: ah ok, I was kind of worried there for a bit. :sweat_smile:

File: client/cli/src/commands/export_sync_state_cmd.rs [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+use sc_service::chain_ops::{MaybeChtRootStorageProvider, build_light_sync_state};
+use structopt::StructOpt;
+use std::io::Write;
+use std::sync::Arc;
+use std::task::Poll;
+use sp_runtime::traits::Block as BlockT;
+
+/// Export the chain spec containing a state to sync the light client.
+#[derive(Debug, StructOpt)]
+pub struct ExportSyncStateCmd {
```
Comment (bkchr) [inline client/cli/src/commands/export_sync_state_cmd.rs:34]: Why is that a new CLI command?

Why isn't that just part of `build-spec` as cli argument? 

I don't really see the reason for a new command.

File: client/cli/src/runner.rs [hunk 10]
```diff
@@ -177,19 +178,25 @@ impl<C: SubstrateCli> Runner<C> {
 
 	/// A helper function that runs a future with tokio and stops if the process receives the signal
 	/// `SIGTERM` or `SIGINT`.
-	pub fn run_subcommand<BU, B, BA, IQ, CL>(self, subcommand: &Subcommand, builder: BU)
-		-> Result<()>
+	pub fn run_subcommand<B, FBA, IQ, FCL, FB, LB, LCL, LBA, SO>(
+		self, subcommand: &Subcommand, full_builder: FB, light_builder: LB,
```
Comment (bkchr) [inline client/cli/src/runner.rs:182]: Can we please find a better solution for this.

There is stuff that doesn't support light client and I don't want to dance around these type declarations again. (Cumulus)

We had this once for `run_node` and it was horrible.

File: client/cli/src/commands/export_sync_state_cmd.rs (hunk 7 above)
Comment (bkchr) [reply client/cli/src/commands/export_sync_state_cmd.rs:68]: The code you have written is essentially just:

```
while sync_oracle.is_major_syncing() || sync_oracle.is_offline() {}
```

Comment (bkchr) [reply client/cli/src/commands/export_sync_state_cmd.rs:68]: Ahh no, not even that. It just calls this future once and than it is stuck, because nothing wakes the future again.

Comment (tomaka) [reply client/cli/src/commands/export_sync_state_cmd.rs:68]: Yeah, I believe you should call `client.import_notification_stream()` and check `is_major_sync() || is_offline()` in response to each newly imported block.
Maybe add a timeout in case we're indeed offline.

Comment (expenses) [reply client/cli/src/commands/export_sync_state_cmd.rs:68]: Using `NetworkStatusSinks` would actually be a better idea, no?

File: bin/node/cli/src/service.rs [hunk 11]
```diff
@@ -427,19 +428,20 @@ pub fn new_light_base(config: Configuration) -> Result<(
...
 		})?;
 
-	Ok((task_manager, rpc_handlers, client, network, transaction_pool))
+	Ok((task_manager, client, backend, rpc_handlers, network, transaction_pool))
 }
 
 /// Builds a new service for a light client.
 pub fn new_light(config: Configuration) -> Result<TaskManager, ServiceError> {
-	new_light_base(config).map(|(task_manager, _, _, _, _)| {
+	new_light_base(config).map(|(task_manager, ..)| {
```
Comment (cecton) [inline bin/node/cli/src/service.rs:444]: :open_mouth: I didn't know

File: client/cli/src/commands/export_sync_state_cmd.rs (hunk 7 above)
Comment (tomaka) [reply client/cli/src/commands/export_sync_state_cmd.rs]: It's not really meant for this kind of usage. The entire `NetworkStatus` is a hack.

Comment (cecton) [reply client/cli/src/commands/export_sync_state_cmd.rs]: It is a... task that wakes up all the time and check the status of something? Why would anyone need something like that. At this rate you can just make a loop with a `sleep(1sec).await` in it :thinking: probably more readable too

File: client/cli/src/commands/build_spec_cmd.rs [hunk 12]
```diff
@@ -75,14 +86,46 @@ impl BuildSpecCmd {
...
+			CL: sp_blockchain::HeaderBackend<B>,
+			BA: MaybeChtRootStorageProvider<B>,
+	{
+		network_status_sinks.network_status(std::time::Duration::from_secs(1)).take_while(|(status, _)| {
+			ready(if status.sync_state == sc_network::SyncState::Idle && status.num_sync_peers > 0 {
+				false
+			} else {
+				true
+			})
+		}).for_each(|_| ready(())).await;
```
Comment (bkchr) [inline client/cli/src/commands/build_spec_cmd.rs:110]:
```suggestion
		network_status_sinks.network_status(std::time::Duration::from_secs(1)).filter(|(status, _)| {
			ready(status.sync_state == sc_network::SyncState::Idle && status.num_sync_peers > 0)
		}).into_future().await;
```

File: client/cli/src/commands/build_spec_cmd.rs [hunk 13]
```diff
@@ -33,6 +39,11 @@ pub struct BuildSpecCmd {
 	#[structopt(long = "raw")]
 	pub raw: bool,
 
+	/// Sync the chain using a light client, and export the state in the chain spec so that other
+	/// light clients can sync faster.
+	#[structopt(long = "export-sync-state")]
```
Comment (bkchr) [inline client/cli/src/commands/build_spec_cmd.rs:44]:
```suggestion
	#[structopt(long)]
```

File: client/cli/src/runner.rs [hunk 14]
```diff
@@ -177,45 +179,61 @@ impl<C: SubstrateCli> Runner<C> {
...
+		LBA: sc_service::chain_ops::MaybeChtRootStorageProvider<B>,
 	{
 		let chain_spec = self.config.chain_spec.cloned_box();
 		let network_config = self.config.network.clone();
 		let db_config = self.config.database.clone();
 
 		match subcommand {
-			Subcommand::BuildSpec(cmd) => cmd.run(chain_spec, network_config),
+			Subcommand::BuildSpec(cmd) => {
+				if cmd.export_sync_state {
```
Comment (bkchr) [inline client/cli/src/runner.rs:207]: No please. Please stop adding hacks over hacks.

File: client/cli/src/commands/build_spec_cmd.rs [hunk 15]
```diff
@@ -75,14 +86,46 @@ impl BuildSpecCmd {
 		}
 		Ok(())
 	}
+
+	/// Sync the light client, export the sync state and run the command as per normal.
+	pub async fn run_export_sync_state<B, CL, BA>(
```
Comment (bkchr) [inline client/cli/src/commands/build_spec_cmd.rs:91]: I still don't like that this only works for light clients...

However, if you still want to follow this path. Make the function return an error when the user tries to run a full node when using this.

File: client/cli/src/commands/build_spec_cmd.rs [hunk 16]
```diff
@@ -75,14 +86,46 @@ impl BuildSpecCmd {
...
+				false
+			} else {
+				true
+			})
+		}).for_each(|_| ready(())).await;
+
+		let light_sync_state = build_light_sync_state(client, backend)?;
+		spec.set_light_sync_state(light_sync_state.to_serializable());
+
+		self.run(spec, network_config)
```
Comment (bkchr) [inline client/cli/src/commands/build_spec_cmd.rs:115]: Did you tested this? 

Why are we running the node here?

File: client/cli/src/runner.rs (hunk 14 above)
Comment (expenses) [reply client/cli/src/runner.rs:207]: If we don't do this, we have to either pass `light_builder` to `cmd.run` or build the light client/backend/etc. even if it's not going to be used.

Comment (cecton) [reply client/cli/src/runner.rs:207]: Actually :thinking: building the light client backend even if it's not used might not be a bad idea. Of course it's not optimal but that just slow down the initialization. If it can improve the situation in term of readability/maintenance, it might be a good trade

File: client/cli/src/commands/build_spec_cmd.rs (hunk 15 above)
Comment (cecton) [reply client/cli/src/commands/build_spec_cmd.rs:91]: Maybe a bad question but isnt the light client a subset of the full client? Can't we export the data with a full client, strip the data of unneeded content, and import it with a light client? So it could maybe work for light and full?

File: client/cli/src/runner.rs [hunk 17]
```diff
@@ -177,45 +179,61 @@ impl<C: SubstrateCli> Runner<C> {
...
+					let (client, backend, network_status_rx, task_manager) = light_builder(self.config)?;
+					run_until_exit(
+						self.tokio_runtime,
+						cmd.run_export_sync_state(chain_spec, network_config, client, backend, network_status_rx),
+						task_manager
+					)
```
Comment (cecton) [inline client/cli/src/runner.rs:213]: I know it's not directly related to this issue but I just noticed the run_until_exit is wrong.

Can you replace, in run_until_exit, this:
```
	task_manager.terminate();
	drop(tokio_runtime);
```

by this:

```
	task_manager.clean_shutdown();
```

File: client/cli/src/commands/build_spec_cmd.rs (hunk 15 above)
Comment (bkchr) [reply client/cli/src/commands/build_spec_cmd.rs:91]: You could build the data also on a full node. That was my point, why I don't like that it is only done on light clients :D

File: client/cli/src/runner.rs (hunk 14 above)
Comment (bkchr) [reply client/cli/src/runner.rs:207]: As I said below, just throw an error when the user started the node as a full node. This is still better than this.

**Comment by expenses** [comment]
Another point for discussion: It would be nice to have a few of the other subcommands work for the light client as well. For example, running `substrate --light` then `substrate revert` fails with `Error: Service(Client(Backend("Unexpected database type. Expected: full")))`.

File: client/cli/src/runner.rs (hunk 17 above)
Comment (expenses) [reply client/cli/src/runner.rs:213]: I occasionally get a few errors on shutdown after this such as
```
2020-08-18 10:24:29.218 tokio-runtime-worker WARN telemetry  Ignored telemetry message because of error on channel: TrySendError { kind: Disconnected }
```

```
Thread 'tokio-runtime-worker' panicked at 'slog-scope: No logger set. Use `slog_scope::set_global_logger` or `slog_scope::scope`.', /home/ashley/.cargo/registry/src/github.com-1ecc6299db9ec823/slog-scope-4.3.0/lib.rs:125
```

Comment (cecton) [reply client/cli/src/runner.rs:213]: It should improve, not the opposite XD

I will make a branch to see if that triggers error in the ci (the ci more easily triggered by wrong shutdown)

Comment (cecton) [reply client/cli/src/runner.rs:213]: #6909

File: client/cli/src/commands/build_spec_cmd.rs [hunk 18]
```diff
@@ -33,6 +39,16 @@ pub struct BuildSpecCmd {
...
 
+	/// Export the light client state in the chain spec so that other light clients can sync faster.
+	/// Generally called together with `--sync-light-client`.
+	#[structopt(long)]
+	pub export_sync_state: bool,
+
+	/// Sync the chain using a light client first. Does nothing unless `--export-sync-state` is also
+	/// called.
+	#[structopt(long)]
+	pub sync_light_client: bool,
```
Comment (bkchr) [inline client/cli/src/commands/build_spec_cmd.rs:50]:
```suggestion
	/// Sync the chain using a light client first.
	#[structopt(long, requires = "export-sync-state")]
	pub sync_light_client: bool,
```

File: bin/node-template/node/src/command.rs [hunk 19]
```diff
@@ -69,11 +69,19 @@ pub fn run() -> sc_cli::Result<()> {
...
+					let PartialComponents { client, backend, task_manager, import_queue, .. }
+						= new_partial(&config)?;
+					Ok((client, backend, import_queue, task_manager))
+				},
+				|config| {
+					let (task_manager, client, backend, network_status_sinks)
+						= new_light_base(config)?;
+					Ok((client, backend, network_status_sinks, task_manager))
+				},
+			)
```
Comment (expenses) [inline bin/node-template/node/src/command.rs:84]: What if we instead did something like this:

```suggestion
      let runner = cli.create_runner(subcommand)?;
			if cli.role == Role::Light {
				runner.run_subcommand(subcommand, |config| {
					let (task_manager, client, backend, import_queue, network_status_sinks, ..) =
						new_light_base(config)?;
					Ok((client, backend, import_queue, network_status_sinks, task_manager))
				});
			} else {
				runner.run_subcommand(subcommand, |config| {
					let PartialComponents { client, backend, task_manager, import_queue, network_status_sinks, ..}
						= new_partial(&config)?;
					Ok((client, backend, import_queue, network_status_sinks, task_manager))
				});
			}
```


Then we'd get rid of the new generics in `run_subcommand` and allow the other subcommands to run for the light client.

Comment (cecton) [reply bin/node-template/node/src/command.rs:84]: Ah no you cant. `cli.role` is the one of runcmd, not the one of your subcommand. I know this thing is really a trap we fixed a similar issue in cumulus recently.

TL;DR: don't use cli..... it's not the command line arguments you would expect

Comment (expenses) [reply bin/node-template/node/src/command.rs:84]: Would it be possible to have the `--light` flag no matter which command we're running?

Comment (cecton) [reply bin/node-template/node/src/command.rs:84]: you can use `runner.config().role` instead

Comment (expenses) [reply bin/node-template/node/src/command.rs:84]: Yeah, but that's set by the subcommand, while we want to be able to run (almost) any subcommand on either a light or full node.

At the moment if I run `cargo run -- --light build-spec` I get:

```
error: Found argument 'build-spec' which wasn't expected, or isn't valid in this context

USAGE:
    substrate --light
```

Comment (bkchr) [reply bin/node-template/node/src/command.rs:84]: We don't support this kind of arguments and also don't want to support them.

Comment (expenses) [reply bin/node-template/node/src/command.rs:84]: I don't really see any way I can make this code cleaner then.

File: client/cli/src/runner.rs [hunk 20]
```diff
@@ -177,45 +178,56 @@ impl<C: SubstrateCli> Runner<C> {
...
+		LBA: sc_service::chain_ops::MaybeChtRootStorageProvider<B>,
 	{
 		let chain_spec = self.config.chain_spec.cloned_box();
 		let network_config = self.config.network.clone();
 		let db_config = self.config.database.clone();
 
 		match subcommand {
-			Subcommand::BuildSpec(cmd) => cmd.run(chain_spec, network_config),
+			Subcommand::BuildSpec(cmd) => {
+				let (client, backend, network_status_rx, task_manager) = light_builder(self.config)?;
```
Comment (bkchr) [inline client/cli/src/runner.rs:206]: The `build-spec` command is now always executed with a light client. This is really really weird behavior...

**Comment by expenses** [comment]
@gnunicorn We're having some major trouble here trying to find a way to get this CLI subcommand to play nice with the others. Most subcommands require a few components (client, backend, import queue) of a partially-built full service. `--build-spec` now (some of the time) requires a few components of a **fully-built** light service. Because of this, `run_subcommand` needs to be passed a bunch of stuff from the construction function of a light service, but there doesn't seem to b

**Comment by expenses** [comment]
Oops

**Comment by expenses** [comment]
Given these struggles that we're facing, it might be a better idea to go back to having a RPC call.

**Comment by cecton** [comment]
If we want the sc-cli commands to support light/full node, then we should be clear about it and first refactor sc-cli to handle this. In the current state it is really not trivial to implement. But as I understand from @bkchr , we do not want that? But maybe that decision can be challenged (thinking of @gnunicorn here as he manages the client in general, but other stakeholders on the task?).

I personally have no opinion about it (I just understand on high level) but there will be a lot of code to make this happen in sc-cli if we do proceed in that direction.

**Comment by expenses** [comment]
I now believe that this it would be a lot easier if the behaviour we're going for (running a light client as normal, but then stopping once synced and outputting a chain spec) was part of the run command instead. We'd add 2 flags to `RunCmd` that require `--light`, `--export-sync-state` (which on it's own would stop the node immediately). and `--sync-before-exporting`, which would make it sync first.

I'm not entirely sure how we'd implement this, perhaps by modifying the `main` function in `sc-cli` to take a future that takes an `exit_future::Exit` that fires when the node has synced?

**Comment by expenses** [comment]
We could also have a different kind of subcommand that are a lot closer in how they operate to the `run` command than other subcommands.

File: client/rpc-api/src/system/mod.rs (hunk 5 above)
Comment (rphmeier) [reply client/rpc-api/src/system/mod.rs]: > So it sounds like generating this data on the full node will either take up storage space or cpu time. I think it would be better to just do this on the light client for now.

Can you expand on what you mean here? We want to generate the data on a full node, and use it on light nodes

Comment (expenses) [reply client/rpc-api/src/system/mod.rs]: @rphmeier but what's the benefit in that over just gathering it on a light node?

**Comment by cecton** [comment]
You mean something like `./substrate run --light --stop-after-sync`? I think it is sounds... at least it will be easier to implement.

> I'm not entirely sure how we'd implement this, perhaps by modifying the main function in sc-cli to take a future that takes an exit_future::Exit that fires when the node has synced?

Maybe not in sc-cli but in sc-service instead.

I think you would need to add a parameter in sc_service::Configuration: "stop_after_sync: bool". When the node is started, if the parameter is `true`, the node would stop when the sync is completed. That would make this parameter available for full and for light client I think. (I don't know if it is sounds to have this feature for the full client too.)

In the TaskManager you will probably need to add something that would allow the exit you were talking about.

Small note about `exit_future` vs channels: it might make send to use a channel instead (the Signal is not clonable VS sender's of channels are clonable). It depends on what you are going to implement.

Few ideas:
- make the SpawnEssentialTask takes futures that would return a boolean (instead of `()`). A `false` return value would mean the node needs to be terminated
- make another SpawnSomethingTask that would end the task manager when the task completes

To have a more tidy implementation. I would suggest to port the loop you did with the check to sc-service or sc-network (dont remember). There is another loop somewhere that is already async that could do the check every time the block is imported. The end goal would be to have something async that we can await and when it ends it means the sync is done. That will look much better than the `loop { poll }` (I was not a big fan! sorry :sweat_smile:).

Overall implementing a "future-exiting-task-manager" feature like that would be a great addition because people would be able to make their own future that would exit the node on some condition.

**Comment by cecton** [comment]
(Comes to think of it... :thinking:  I wonder if `fn future()` in the taskmanager and `fn clean_shutdown()` couldn't merged into a single function :thinking: there isn't much point to have them both separated)

File: client/rpc-api/src/system/mod.rs (hunk 5 above)
Comment (cecton) [reply client/rpc-api/src/system/mod.rs]: I think the idea is that you would be able to generate the data from a node who is already populated (typically a full node) and then give to a machine that wants to run a light node from those data (or at least a subset of it).

**Comment by bkchr** [comment]
Okay, I think the solution for this pr is the following. You revert the changes on the CLI for now. We merge this pr.

After this, you create a new pr. In this pr we drop the default support for subcommands. This means people will need to choose what kind of sub-commands they want. And if we do this, it will be much more easier to provide the correct client etc to the sub-command that should be executed.

Substrate is too extensible that it makes more sense for people to choose the sub commands "manually" when implementing their nodes.

File: bin/node/cli/src/service.rs (hunk 11 above)
Comment (seunlanlege) [reply bin/node/cli/src/service.rs:445]: woah :exploding_head:

**Comment by expenses** [comment]
@bkchr good idea. I've reverted those changes for now.

File: client/service/src/chain_ops/build_spec.rs [hunk 21]
```diff
@@ -0,0 +1,82 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
```
Comment (expenses) [inline client/service/src/chain_ops/build_spec.rs:1]: I'm not sure if I should move the contents of this file or not.

**Review by bkchr — APPROVED** [review]

File: client/service/src/chain_ops/build_spec.rs (hunk 21 above)
Comment (bkchr) [reply client/service/src/chain_ops/build_spec.rs:1]: I think it is fine for now

**Comment by bkchr** [comment]
> @bkchr good idea. I've reverted those changes for now.

Thank you

**Comment by cecton** [comment]
:thinking: I guess that makes sense. We remove the subommand enum then?

**Comment by bkchr** [comment]
> thinking I guess that makes sense. We remove the subommand enum then?

Yes

**Comment by cecton** [comment]
> > thinking I guess that makes sense. We remove the subommand enum then?
> 
> Yes

@expenses don't forget to remove the macro then

**Comment by bkchr** [comment]
> > > thinking I guess that makes sense. We remove the subommand enum then?
> > 
> > 
> > Yes
> 
> @expenses don't forget to remove the macro then

Let's wait for the pr ;)

**Comment by expenses** [comment]
bot merge
